handout/testing/Introduction to TDD Testing.md

```diff
@@ -0,0 +1,185 @@
+
+In order to test our Angular 2 application we need create an environment for it to run in. We could use a browser like Chrome or Firefox to accomplish this (Karma supports in-browser testing). We could also use a browser-less environment to test our application, which can offer us greater control over automating certain tasks and managing our testing workflow. [Phantom-js](http://phantomjs.org/) provides a javascript api that allows us to create a headless DOM instance which can be used to bootstrap our Angular 2 application. Then, using that DOM instance that is running our Angular 2 application, we can run our tests. 
+
+[Istanbul](https://gotwarlost.github.io/istanbul/) is used by Karma to generate code coverage reports, which will tell us to what overall percent our application is being tested. This is a great way to track what components/services/pipes/etc have tests written and which don't. We can get some useful insight into how much are application is being tested and where. 
+
+For some extra testing functionality we can use the [Sinon](http://sinonjs.org/) library for things like test spys, test subs, and mock XHR requests. 
+
+[Chia](http://chaijs.com/) is an assertion library that can be paired with any other testing framework. It offers some sytnax sugar that lets us write our unit tests with different verbage - we can use a *should*, *expect* or *assert* interface. Chia also takes advantage of function chaining to form english like sentences used to describe tests in a more user friendly way. Chia isn't a required library for testing and we won't explore it much more in this handout, but it is a useful tool for creating cleaner more well written tests. 
+
+## Setup
+
+The repo [ng2-redux-starter](https://github.com/rangle/angular2-redux-starter) is a basic webpack based Angular 2 application (with Redux) with the same testing toolchain outlined above. Lets take a look at how this project is setup. 
+
+###Filename Conventions
+Each unit test is put into its own separate file. The Angular 2 team recommends putting unit test scripts alongside the files they are testing and using a `.spec` filename extension to mark it as a testing script (this is a Jasmine convention). So if you had a component `/app/components/MyComponent.ts`, then your unit test for this component would be in `/app/components/MyComponent.spec.ts`. This is a matter of taste, you can put your testing scripts wherever you like, though keeping them close to your source files makes them easier to find and gives those who aren't familiar with the source code an idea of how that particular piece of code should work. 
```

Comment:
I suggest we use all lowercase `mycomponent.ts` as this is our code convention anyway